plugins/basic/nexus-repository-maven/src/main/java/org/sonatype/nexus/repository/maven/internal/DigestExtractor.java

```diff
@@ -0,0 +1,159 @@
+   * Extract the digest string from a blob, that carries content of a Maven sha1/md5 file.
+   *
+   * @see {@link #extract(String)}
+   */
+  @Nullable
+  public static String extract(final Blob blob)
+      throws IOException
+  {
+    checkNotNull(blob);
+    try (InputStream is = blob.getInputStream()) {
+      return extract(is);
+    }
+  }
+
+  /**
+   * Extract the digest string from a stream, that carries a payload of a Maven sha1/md5 file.
+   *
+   * @see {@link #extract(String)}
+   */
+  @Nullable
+  public static String extract(final InputStream stream)
+      throws IOException
+  {
+    checkNotNull(stream);
+    try (InputStreamReader isr = new InputStreamReader(
+        new LimitedInputStream(stream, 0, MAX_CHARS_NEEDED), Charsets.UTF_8)) {
+      return extract(CharStreams.toString(isr));
+    }
+  }
+
+  /**
+   * Extract the checksum from a string. If result appears to be a digest (cannot be 100% sure), it is returned,
+   * otherwise {@code null}.
+   */
+  @Nullable
+  public static String extract(final String input) {
+    checkNotNull(input);
+    // we need most of MAX_CHARS_NEEDED characters from input
+    String raw = input.length() > MAX_CHARS_NEEDED ? input.substring(0, MAX_CHARS_NEEDED) : input;
+    // if multiline, we need the 1st line only
+    raw = raw.indexOf('\n') > -1 ? raw.substring(0, raw.indexOf('\n')) : raw;
+    // trim whatever we have
+    raw = raw.trim();
+
+    // check did we end up with empty string
+    if (Strings.isNullOrEmpty(raw)) {
+      return null;
+    }
+
+    String digest;
+    // digest string at end with separator, e.g.:
+    // MD5 (pom.xml) = 68da13206e9dcce2db9ec45a9f7acd52
+    // ant-1.5.jar: DCAB 88FC 2A04 3C24 79A6 DE67 6A2F 8179 E9EA 2167
+    if (raw.contains("=") || raw.contains(":")) {
+      digest = raw.split("[=:]", 2)[1].trim();
+    }
+    else {
+      // digest string at start, e.g. '68da13206e9dcce2db9ec45a9f7acd52 pom.xml'
+      digest = raw.split(" ", 2)[0];
+    }
+
+    if (!isDigest(digest)) {
+      // maybe it's "uncompressed", e.g. 'DCAB 88FC 2A04 3C24 79A6 DE67 6A2F 8179 E9EA 2167'
+      digest = compress(digest);
+    }
+
+    if (!isDigest(digest)) {
+      // check if the raw string is an uncompressed checksum, e.g.
+      // 'DCAB 88FC 2A04 3C24 79A6 DE67 6A2F 8179 E9EA 2167'
+      digest = compress(raw);
+    }
+
+    if (!isDigest(digest) && raw.contains(" ")) {
+      // check if the raw string is an uncompressed checksum with file name suffix, e.g.
+      // 'DCAB 88FC 2A04 3C24 79A6 DE67 6A2F 8179 E9EA 2167 pom.xml'
+      digest = compress(raw.substring(0, raw.lastIndexOf(" ")).trim());
+    }
+
+    if (isDigest(digest)) {
+      return digest;
+    }
+    else {
+      return null;
+    }
+  }
+
+  /**
+   * Returns {@code true} if the input string "looks like" digest hex (md5 or sha1).
+   */
+  public static boolean isDigest(final String digest) {
+    return digest.length() >= 32 && digest.matches("^[a-z0-9]+$");
+  }
+
+  // ==
```

Comment:
Please stop putting in these markers.  If you want to denote sections, then give them details.  I've found so many places where code is littered it "// ==" markers even ever other method in some places.  Its meaningless noise.

I don't object to making logical sections in code to organize related things, but name them and keep them updated.  With "//  ==" having no real meaning or context its var to easy for the organization you are indending to degrade very  quickly.